Sources/_MatchingEngine/Regex/Parse/LexicalAnalysis.swift

```diff
@@ -620,97 +623,116 @@ extension Source {
-          return .changeMatchingOptions(seq, isIsolated: true)
-        }
 
-        guard let next = src.peek() else {
-          throw ParseError.expectedGroupSpecifier
+          // Matching option changing group (?iJmnsUxxxDPSWy{..}-iJmnsUxxxDPSW:).
+          if let seq = try src.lexMatchingOptionSequence() {
+            if src.tryEat(":") {
+              return .changeMatchingOptions(seq, isIsolated: false)
+            }
+            // If this isn't start of an explicit group, we should have an
+            // implicit group that covers the remaining elements of the current
+            // group.
+            // TODO: This implicit scoping behavior matches Oniguruma, but PCRE
+            // also does it across alternations, which will require additional
+            // handling.
+            guard src.tryEat(")") else {
+              if let next = src.peek() {
+                throw ParseError.invalidMatchingOption(next)
+              }
+              throw ParseError.expected(")")
+            }
+            return .changeMatchingOptions(seq, isIsolated: true)
+          }
+
+          guard let next = src.peek() else {
+            throw ParseError.expectedGroupSpecifier
+          }
+          throw ParseError.unknownGroupKind("?\(next)")
         }
-        throw ParseError.misc("Unknown group kind '(?\(next)'")
-      }
 
-      // Explicitly spelled out PRCE2 syntax for some groups.
-      if src.tryEat("*") {
-        if src.tryEat(sequence: "atomic:") { return .atomicNonCapturing }
+        // Explicitly spelled out PRCE2 syntax for some groups.
+        if src.tryEat("*") {
+          if src.tryEat(sequence: "atomic:") { return .atomicNonCapturing }
 
-        if src.tryEat(sequence: "pla:") ||
-            src.tryEat(sequence: "positive_lookahead:") {
-          return .lookahead
-        }
-        if src.tryEat(sequence: "nla:") ||
-            src.tryEat(sequence: "negative_lookahead:") {
-          return .negativeLookahead
-        }
-        if src.tryEat(sequence: "plb:") ||
-            src.tryEat(sequence: "positive_lookbehind:") {
-          return .lookbehind
-        }
-        if src.tryEat(sequence: "nlb:") ||
-            src.tryEat(sequence: "negative_lookbehind:") {
-          return .negativeLookbehind
-        }
-        if src.tryEat(sequence: "napla:") ||
-            src.tryEat(sequence: "non_atomic_positive_lookahead:") {
-          return .nonAtomicLookahead
-        }
-        if src.tryEat(sequence: "naplb:") ||
-            src.tryEat(sequence: "non_atomic_positive_lookbehind:") {
-          return .nonAtomicLookbehind
-        }
-        if src.tryEat(sequence: "sr:") || src.tryEat(sequence: "script_run:") {
-          return .scriptRun
-        }
-        if src.tryEat(sequence: "asr:") ||
-            src.tryEat(sequence: "atomic_script_run:") {
-          return .atomicScriptRun
+          if src.tryEat(sequence: "pla:") ||
+              src.tryEat(sequence: "positive_lookahead:") {
```

Comment:
Yeah that would be nice, I also kinda feel like we should drop the collection variant of `tryEat(anyOf:)` and then drop the `sequence:` label in favor of just having `tryEat(_:)` and variadic `tryEat(anyOf:)` that works for both strings and characters (or even just overloads that take a string argument). What do you think?

Reply:
I've gone down the rabbit hole a couple of times, sometimes it works out and sometimes it doesn't. The argument label can be useful or even essential, since string literals can represent both string _and_ characters. It is true that the semantics are often the same, but another common API to want is an `anyOf` that takes a collection of Character, which String also satisfies.

You can explore it, but in general I've found it to be cleaner if the `Element` eat and `Sequence` eat have different names somehow.

Reply:
> `anyOf` that takes a collection of Character, which String also satisfies

For that I'd probably rather have overloads that take e.g an array of characters and set of characters, if you want to pass a string you'd do `Array(str)` to make it clear you want it treated as a collection.